### DISCUSSION TITLE
n00b question about offset commit

### DISCUSSION THREAD

**Discussion by funkiestj** [confluentinc/confluent-kafka-go#1096] [Q&A]
I'm baffled as to why[ CommitMessage()](https://github.com/confluentinc/confluent-kafka-go/blob/5f88a1c61f6588c399428db7ee58402eeab9009a/kafka/consumer.go#L313) is incrementing the offset provided by the message.   

I tried googling to find more detail on the semantics of the committed offset but I can't find it.

**Comment by adilburaksen** [comment]
`CommitMessage()` increments the offset because that's how Kafka offset semantics work — the committed offset always represents the **next** offset to consume, not the last consumed one.

If you consumed message at offset 5, you commit offset 6. This means: "if this consumer restarts, start from offset 6 (the next unconsumed message)." Committing offset 5 would cause the message at offset 5 to be reprocessed on restart.

This is standard Kafka behavior, not a library quirk. From the Apache Kafka docs:

> The committed position is the last offset that has been stored securely. Should the process fail and restart, this is the offset that the consumer will recover to.

The offset that's "safely committed" is the _next_ offset to read, so committing N means "I've successfully processed everything up to and including N-1."

```
[offset 4] [offset 5] [offset 6] ...
                ↑
         You consumed this.
         CommitMessage commits offset 6.
         On restart: consume from 6 (not 5).
```

This `+1` increment prevents re-processing the message you just committed, which is what you want for at-least-once delivery semantics.
